### ISSUE TITLE
`apply_updates` Causes Unnecessary Dtype Conversions Leading to Precision Loss

### ISSUE THREAD

**Issue by Aaryan-549** [google-deepmind/optax#1444]
The current implementation of `optax.apply_updates()` performs an unnecessary dtype conversion that can cause precision loss and additional computational overhead in mixed-precision training scenarios.

**Current implementation** (from `optax/_src/update.py`):
```python
def apply_updates(params: base.Params, updates: base.Updates) -> base.Params:
  return jax.tree.map(
      lambda p, u: (
          None if p is None else jnp.asarray(p + u).astype(jnp.asarray(p).dtype)
      ),
      params,
      updates,
      is_leaf=lambda x: x is None,
  )
```

## Problem

The function performs the operation in this order:
1. Convert `p` to array: `jnp.asarray(p)`
2. Add `p + u` (this may promote dtypes)
3. Convert result to array: `jnp.asarray(p + u)` 
4. Cast back to original dtype: `.astype(jnp.asarray(p).dtype)`

This causes **two issues**:

1. **Precision Loss**: When `updates` have higher precision than `params` (e.g., params in `float16`, updates in `float32` from gradient accumulation), the sum `p + u` is computed in higher precision but then cast down, losing the precision benefit.

2. **Redundant Computation**: The double `jnp.asarray()` call is redundant since `jnp.astype()` already returns an array.

## Expected Behavior

The function should:
- Compute the sum in the appropriate precision
- Preserve parameter dtype without unnecessary intermediate conversions
- Avoid redundant array conversions

## Proposed Solution

```python
def apply_updates(params: base.Params, updates: base.Updates) -> base.Params:
  def _apply_update(p, u):
    if p is None:
      return None
    # Compute in higher precision if needed, then cast once
    p_array = jnp.asarray(p)
    result = p_array + u
    # Only cast if dtypes differ (avoids no-op casts)
    if result.dtype != p_array.dtype:
      return result.astype(p_array.dtype)
    return result
    
  return jax.tree.map(
      _apply_update,
      params,
      updates,
      is_leaf=lambda x: x is None,
  )
```

### Step 1: Create comprehensive test cases
- Test with mixed precision (float16 params, float32 updates)
- Test with matching dtypes
- Test with None values in pytree
- Test with complex dtypes
- Benchmark performance impact

### Step 2: Implement the fix
- Refactor `apply_updates` to avoid redundant conversions
- Ensure backward compatibility

### Step 3: Validate
- Run existing test suite
- Add new tests for precision preservation
- Benchmark against current implementation

### Step 4: Documentation
- Update docstring to clarify dtype handling behavior
- Add example showing mixed-precision use case

## Impact

- **Users affected**: Anyone using mixed-precision training or working with different dtypes between params and updates
- **Benefit**: Better numerical precision, slight performance improvement
- **Risk**: Low - maintains same API and behavior for common cases

---

This is a concrete, solvable issue with clear scope and measurable impact. Would you like me to help you implement this fix?

**Comment by rdyro** [comment]
In mixed precision training, both accumulating in potentially higher precision of updates and retaining the original dtype of params is expected. The current implementation is correct.

This action plan is quite ambitious, I'd recommend splitting a proposed contribution into smaller pieces.

**Comment by Aaryan-549** [comment]
That's great, thank you! I'm excited to contribute and will start working on the PR right away.

**Comment by rdyro** [comment]
The current library implementation is correct, this is not an issue I'd recommend taking up.

**Comment by Aaryan-549** [comment]
Thank you for the clarification on that. I appreciate you taking the time to review it.

I've also opened a separate issue regarding input validation, which I believe is a valid point to address: #1447.

I'd be grateful if you could have a look at that one when you have a moment. I'm still very keen to find a way to contribute.
